bindings/python/pinocchio/display/abstract_display.py

```diff
@@ -0,0 +1,70 @@
+from .. import libpinocchio_pywrap as pin
+from ..shortcuts import buildModelsFromUrdf, createDatas
+
+import time
+
+class AbstractDisplay(object):
+    """Pinocchio displays are employed to easily display a model at a given configuration.
+    AbstractDisplay is not meant to be directly employed, but only to provide a uniform interface and a few common methods.
+    New displays should extend this class and override its methods as neeeded.
+    """
+
+    @classmethod
+    def BuildFromURDF(cls,filename, package_dirs=None, root_joint=None, verbose=False, meshLoader=None):
```

Comment:
This method is really specific to Gepetto viewer and should be set as a common method to all the viewers.
In addition, this option is in Gepetto viewer for historical reason, but Pinocchio allows to bypass it easily.

Reply:
Why is it specific to Gepetto viewer? All viewer will need the model, the visual model, and the collision model. You mean it should *not* be set as a common method? Why? It's a convenient way to build the viewer.

Reply:
Because Gepetto Viewer comes with an internal URDF parser, which allows the direct loading of the geometries. But this is not the case of all the viewers we target, and a simple way to a load a model inside a viewer is simply by looking to the meshes path stored inside the Geometries. This is the main way to achieve it, without requiring urdfdom and others to be connected to your viewer.

Reply:
In addition, a viewer does not that there is a kinematic tree behind the scene. Its main functionnality is to display meshes.

Reply:
Either there is a misunderstanding or I just do not follow you.

This method is just to instantiate the Pinocchio `Model` and the two `GeometryModel` objects.
If they are stored in a URDF, you need to read them, no matter the viewer.

The loading of the meshes inside the viewer is then handled by `loadDisplayModel`. This method is specialized for each viewer.

> a viewer does not know that there is a kinematic tree behind the scene. Its main purpose is to display meshes.

The viewer does not, but the Display does. Method `display(q)` calls `forwardKinematics` and `updateGeometryPlacements` in order to compute the placement of each mesh for the given joint configuration `q`, and then passes the resulting placements to the viewer. You need knowledge about the models for that.

Otherwise, the `GeometryData` objects should be updated externally. It would then make sense to share them with external code. But I thought it was preferred to keep them separate.

Reply:
Ok. I got your point.
In this case, I do not this worthy to add these two methods at the viewer level.

The ideal and normal usage case is:
1. load or create your Pinocchio model
2. display it calling the default constructor of the Viewer taking as arguments the various models

I would be afraid that adding the URDF methods will lead to bad habits at the user level, with some questions about the sharing of the model for instance. And from a conceptual point of view again, a generic Viewer should not be aware of the URDF convention and so on. It should only take care of the meshes or the basic geometries, nothing more. We already discussed all these points a while ago with @jmirabel, @nmansard, @olivier-stasse, @florent-lamiraux to name a few.

Reply:
> I would be afraid that adding the URDF methods will lead to bad habits at the user level, with some questions about the sharing of the model for instance.

Uhm, I see what you mean. This encourages loading the models through the viewer and then keeping using the models from inside the viewer. Then, I may remove this functionalities so that usage of the dedicated tools (`buildModelsFromUrfs` and `createDatas`) is enforced.

Reply:
Yes, this is exactly what I mean.